File: src/main/java/me/devsnox/battlekits/BattleKits.java
package me.devsnox.battlekits;

import fr.minuskube.inv.InventoryManager;
import me.devsnox.battlekits.commands.*;
import me.devsnox.battlekits.listeners.FrameListener;
import me.devsnox.battlekits.listeners.KitListener;
import me.devsnox.battlekits.listeners.PlayerListener;
import me.devsnox.battlekits.listeners.ShardListener;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;

public class BattleKits extends JavaPlugin {

    private InventoryManager inventoryManager;
    private KitManager kitManager;

    @Override
    public void onEnable() {
        this.inventoryManager = new InventoryManager(this);
        this.inventoryManager.init();

        this.kitManager = new KitManager(this);
        this.kitManager.enable();

        Bukkit.getPluginManager().registerEvents(new KitListener(this.kitManager), this);
        Bukkit.getPluginManager().registerEvents(new PlayerListener(this.kitManager), this);
        Bukkit.getPluginManager().registerEvents(new ShardListener(this.kitManager), this);

        final FrameListener frameListener = new FrameListener();

        this.kitManager.setFrameListener(frameListener);

        Bukkit.getPluginManager().registerEvents(frameListener, this);

        this.getCommand("kiteditor").setExecutor(new KitEditorCommand(this.kitManager));
        this.getCommand("kits").setExecutor(new KitCommand(this.kitManager));
        this.getCommand("getshard").setExecutor(new ShardCommand(this.kitManager));
        this.getCommand("getbooster").setExecutor(new BoosterCommand(this.kitManager));
        this.getCommand("preview").setExecutor(new PreviewTest());
    }

    @Override
    public void onDisable() {
        this.kitManager.disable();
    }
}

File: spigot/src/main/kotlin/me/devsnox/battlekits/BattleKits.kt
package me.devsnox.battlekits

import com.google.gson.JsonObject
import kotlinx.coroutines.runBlocking
import me.devsnox.battlekits.commands.BoosterCommand
import me.devsnox.battlekits.commands.KitCommand
import me.devsnox.battlekits.commands.KitEditorCommand
import me.devsnox.battlekits.commands.ShardCommand
import me.devsnox.battlekits.configs.SQLStatements
import me.devsnox.battlekits.configs.sqlStatements
import me.devsnox.battlekits.events.*
import net.darkdevelopers.darkbedrock.darkness.general.configs.toConfigMap
import net.darkdevelopers.darkbedrock.darkness.general.functions.load
import net.darkdevelopers.darkbedrock.darkness.general.functions.save
import net.darkdevelopers.darkbedrock.darkness.general.functions.toConfigData
import net.darkdevelopers.darkbedrock.darkness.general.functions.toMap
import org.bukkit.plugin.java.JavaPlugin

@Suppress("unused")
class BattleKits : JavaPlugin() {

    private val kitManager by lazy { KitManager(this) }
    private val events by lazy {
        setOf<Setupable>(
            KitEventsTemplate(kitManager),
            PlayerEventsTemplate(kitManager),
            ShardEventsTemplate(kitManager),
            FrameEventsTemplate
        )
    }

    override fun onEnable() {

        val configData = dataFolder.toConfigData("sql-statements", prefix = "DOT_NOT_CHANGE_THIS_CONFIG.")
        sqlStatements = SQLStatements(configData.load<JsonObject>().toMap())
        configData.save(sqlStatements.toConfigMap())

        runBlocking { kitManager.enable() }
        initEvents()
        initCommands()
    }

    override fun onDisable() {
        runBlocking { kitManager.disable() }
        FrameEventsTemplate.reset()
    }

    private fun initEvents(): Unit = events.forEach { it.setup(this) }

    private fun initCommands() {
        BoosterCommand(kitManager)
        KitCommand(kitManager)
        KitEditorCommand(kitManager)
        ShardCommand(kitManager)
    }
}
